keymanapp/keyman
# refactor(web): prep work for mouse-based use of touch events 👆

This starts the next major push of the OSK refactor, continuing from the 📐 chain that ended with #5462.

By adding a layer of abstraction between the touch-event handlers and the methods that actually process the touches, we can prepare a way to have _mouse-based_ interaction use them as well.  This will be very useful for keyboard developers using Developer, as it'll facilitate _much_ simpler swapping among test-device configurations during tests while in development.

The new `inputEventEngine.ts` abstract class will serve as a common abstraction for the mouse event handlers and the touch event handlers.  It'll take some work to make the two conform to the same pattern, as some mouse events don't exactly parallel touch events in their behavior, but I've determined that it's a reasonably tractable approach.  This PR also implements the touch-based version of the abstraction.

Considering our intended design for the OSK in the future, I also put in work to generalize the `VisualKeyboard`'s "event out-of-bounds" detection system while preserving the original intent behind its design.  The "fuzzy bounds" part was quite tricky to work out but also quite necessary - on physical phones and tablets, it's literally impossible to reach a coordinate with `x < 0` in order to cancel touches, which is what the "fuzzy bounds" behavior addresses.
- The critical observation:  what matters is that there is that one or more of the OSK's bounds _within the viewport_ lie too close to the corresponding _screen_ boundary.
    - If there's not sufficient space to run a touchpoint past one of the OSK's boundaries, that boundary can't be used to cancel the touch event.
- From there, a more general approach is possible that covers behavior when anchored to _any_ edge of the window... _and_ also when not even anchored.
    - Though... these claims will require the mouse-based implementation of the new abstraction for validation.  But, this PR's big enough as it is.

Worth noting:  with the "towards the top" shortcut for popup-key cancellation now working again, I consider these changes to be a feature.  Previously, "toward the top" was the only 'safe' path toward cancellation in my experience, which naturally gets noisy if that attempt auto-summons a subkey menu.  Other directions (made available here) won't cause such 'noise'.

------

For this PR, all should appear to function for touch form factors as it always has without changes, save for a few small details:
- It's now possible to 'cancel' a touch event by dragging the touchpoint to the bottom of the screen in a manner similar to dragging it to the left-hand or right-hand sides while still within the OSK.
    - This form of cancellation will occur both _when moving_ and when releasing the touch-point.
        - Previously, this "fuzzy bounds" behavior only worked for the latter.  Which seemed a bit odd.
        - Also, it was never for the OSK's bottom; just the horizontal sides.
    - Only works when moving to a boundary sufficiently far from the initial touch-point.  (Same as before.)
- These details are more pronounced during Chrome emulation, as said emulation _itself_ uses _far_ more generous 'fuzzy bounds' and can report coordinates outside of the emulated screen.  (Only for 12 or so pixels' worth, but it's definitely not 0.)
    - Thus, it's actually possible to verify the behavior during emulation, as cancellation will be "snappier" than Chrome's default before these changes.

These differences result partly from centralizing bounds-checking logic that resided within `moveOver` (for "off the top" bounds) and within `release` (for left and right bounds) and partly from generalizing it to be future-proof when other anchoring / layout behaviors exist than are currently permissible.

------

## User Testing

@keymanapp/testers 

As this PR changes the code responsible for handling user touch interactions with the keyboard, make sure that the standard user experience feels either unchanged or improved if possible.  If there _are_ improvements, please explicitly document these.  Outside of a few notes earlier in this description, things _should_ work the same as before.

**Platform / browser combinations**:
- [ ] Chrome emulation / iOS ("**Prediction - robust testing**" test page)
    - <details><summary>Be sure to enable the Rendering tab's "Emulate a focused device setting."</summary>
    
        <img width="331" alt="image" src="https://user-images.githubusercontent.com/25213402/128671154-f1bd27f2-38d0-4eb3-8080-4083843b76ae.png">

        </details>

- [ ] Android / System keyboard mode
    - Ensure that predictive text is enabled for English.
- [ ] iOS / Keyman app
    - Ensure that predictive text is enabled for English.

<details>
  <summary>Tests (with inlined instructions)</summary>

- [ ] TEST.MOBILE.ROTATE.P-TO-L:
    1. With Keyman not active, load it in a portrait orientation.
    2. Once loaded, select `sil_euro_latin` and longpress the `a` key, selecting the `ä` key.  Confirm that `ä` is emitted.
    3. Then, rotate the device to a landscape orientation.
    4. Again, longpress the `a` key and select the `ä` key.  Confirm that `ä` is emitted.
        - The test "passes" if everything behaves as expected.
- [ ] TEST.MOBILE.ROTATE.L-TO-P:
    1. With Keyman not active, load it in a landscape orientation.
    2. Once loaded, select `sil_euro_latin` and longpress the `a` key, selecting the `ä` key.  Confirm that `ä` is emitted.
    3. Rotate the device to a portrait orientation.
    4. Again, longpress the `a` key and select the `ä` key.  Confirm that `ä` is emitted.
        - The test "passes" if everything behaves as expected.
- [ ] TEST.MOBILE.LAYERS.CHANGE:  Using the `sil_euro_latin` keyboard, swap among the 'default', 'shift', 'numeric', 'symbol', and 'currency' layers.
    - Expected result:  the layer should always change correctly.
    - Expected result:  there should be no "stuck highlighting" effect on the default layer's numeric key.
- [ ] TEST.MOBILE.SPACEBAR_CAP:  Confirm that there is a visible spacebar caption that matches the currently-loaded keyboard.
- [ ] TEST.MOBILE.KEY_PREVIEWS:  On `sil_euro_latin`, press and hold the `o` key.
    - Confirm that the key preview displays momentarily at the correct position (with a larger `o` directly over the key)
- [ ] TEST.MOBILE.SUBKEYS:
    - Confirm that the subkey menu displays properly once the key preview disappears.
    - Confirm that subkey selection works correctly - that the key directly under the touchpoint is selected.
        - Move the touchpoint to the `ȍ` subkey but do not release it.
        - Then, move it to the `ɔ` subkey, _then_ release it.
        - Confirm that a `ɔ` is output.

</details>
